File: internal/model/spaceship.go
package model

import "github.com/google/uuid"

type Spaceship struct {
	ID          uuid.UUID `gorm:"type:uuid;default:gen_random_uuid();primaryKey"`
	Name        string    `gorm:"not null"`
	AstralID    uuid.UUID
	Location    string
	Flight      FlightInfo
	FlightID    uuid.UUID `gorm:"default:null"`
	SystemID    string
	System      System
	PlanetID    string `gorm:"default:null;"`
	Planet      Planet `gorm:"constraint:OnUpdate:CASCADE,OnDelete:SET NULL;"`
	PlayerSitIn *bool  `gorm:"default:false"`
}

File: voidspace/database/models/spaceship.py
from uuid import UUID, uuid4

from sqlalchemy import ForeignKey
from sqlalchemy.orm import Mapped, mapped_column, relationship

from voidspace.database import Base
from .character import Character
from .planet import Planet
from .system import System


class Spaceship(Base):
    __tablename__ = "spaceships"
    id: Mapped[UUID] = mapped_column(primary_key=True, default=uuid4)
    name: Mapped[str]
    location: Mapped[str]
    character: Mapped[Character] = relationship()
    character_id: Mapped[UUID] = mapped_column(ForeignKey("characters.id"))
    active: Mapped[bool]
    system: Mapped[System] = relationship()
    system_id: Mapped[str] = mapped_column(ForeignKey("systems.id"))
    planet: Mapped[Planet] = relationship()
    planet_id: Mapped[str] = mapped_column(
        ForeignKey("planets.id", ondelete="SET NULL", onupdate="CASCADE"), nullable=True
    )
